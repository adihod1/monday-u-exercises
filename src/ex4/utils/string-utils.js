export function capitalizeText(text) {
    return text
        .split(' ')
        .map((s) => s[0].toUpperCase() + s.slice(1))
        .join(' ');
}


export function generateUniqueID() { 
    return Math.random().toString(36).substr(2, 9);
}