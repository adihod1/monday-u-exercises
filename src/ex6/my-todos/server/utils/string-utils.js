
function capitalizeText(text) {
    return text
        .split(' ')
        .map((s) => s[0].toUpperCase() + s.slice(1))
        .join(' ');
}

function generateUniqueID() {
    return Math.random().toString(36).substr(2, 9);
}

module.exports = {
    capitalizeText,
    generateUniqueID
}