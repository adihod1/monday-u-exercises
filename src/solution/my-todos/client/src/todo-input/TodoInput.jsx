import React, { useState, useEffect } from "react";
import "./TodoInput.scss";
import { faPlus } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import ItemClient from "../api/itemClient";
import PropTypes from 'prop-types'

const itemClient = new ItemClient();

let isAddingItem = false;

const addTodo = async (value) => {
    const todos = await itemClient.addTodo(value);
    _duplicatesTodos(todos.body);
}

const getClassname = (active) => ({
    className: active ? 'active' : ''
});

const _duplicatesTodos = (todos) => {
    const duplicates = todos.filter(todo => todo.type == 'pokemonExists' || todo.type == 'todoExists');
    if (!duplicates.length) {
        return;
    }
    let content = 'The following todos are already exist: \n';
    duplicates.forEach((todo) => {
        if(todo.type=='pokemonExists'){
            content += `id: ${todo.pokemon.id}, name: ${todo.pokemon.name} \n`;
        }else{
            content += `${todo.item}\n`
        }
    });
    alert(content);
}

const TodoInput = ({ refreshState, setSpinner }) => {
    const [input, setInput] = useState('');
    const [active, setActive] = useState(false);

    const onChange = (e) => {
        const value = e.target.value;
        setInput(e.target.value);
        setActive(!!value.length);
    }

    const onClick = async () => {
        if (isAddingItem) {
            return
        }
        isAddingItem = true;
        try {
            setSpinner(true);
            await addTodo(input);
            setInput('');
            setSpinner(false);
            refreshState({});
        } catch (e) {
            console.error(e);
        }
        isAddingItem = false;
    }

    const onKeyUp = (e) => input.length && e.key === 'Enter' ? onClick() : null;

    return (
        <div className="todo-input">
            <input onKeyUp={onKeyUp} onChange={onChange} id="todo-input" type="text" placeholder="Add your new todo" value={input}></input>
            <button onClick={onClick} id="todo-input-btn" {...getClassname(active)}><FontAwesomeIcon icon={faPlus} /></button>
        </div>
    );
};

TodoInput.propTypes = {
    setSpinner: PropTypes.func,
    refreshState: PropTypes.func,
}

export default TodoInput;