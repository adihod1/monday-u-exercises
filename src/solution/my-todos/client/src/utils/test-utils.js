export const todosList = [
    {
        id: 1,
        name: 'do homework',
        type: 'text',
        message: 'do homework',
        status: 0
    },
    {
        id: 2,
        name: 'clean the room',
        type: 'text',
        message: 'clean the room',
        status: 1
    },
    {
        id: 3,
        name: 'call a friend',
        type: 'text',
        message: 'call a friend',
        status: 0
    },
    {
        id: 4,
        name: "hjkh",
        type: "text",
        message: "hjkh",
        status: 0
    },
    {
        id: 5,
        name: "bake a cake by the day after tomorrow",
        type: "text",
        message: "bake a cake by the day after tomorrow",
        status: 0
    },
    {
        id: 6,
        name: "84",
        type: "pokemon",
        message: "Catch #84 Doduo the Normal/Flying type pokemon",
        pokemon: {
            pokemon_id: "84",
            pokemon_name: "doduo",
            pokemon_image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/84.png",
            pokemon_type: "[\"Normal\",\"Flying\"]"
        },
        status: 1
    },
    {
        id: 7,
        name: "4",
        type: "pokemon",
        message: "Catch #4 Charmander the Fire type pokemon",
        pokemon: {
            pokemon_id: "4",
            pokemon_name: "charmander",
            pokemon_image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/4.png",
            pokemon_type: "[\"Fire\"]"
        },
        status: 0
    },
    {
        id: 8,
        name: "25",
        type: "pokemon",
        message: "Catch #25 Pikachu the Electric type pokemon",
        pokemon: {
            pokemon_id: "25",
            pokemon_name: "pikachu",
            pokemon_image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/25.png",
            pokemon_type: "[\"Electric\"]"
        },
        status: 0
    },
    {
        id: 9,
        name: "1",
        type: "pokemon",
        message: "Catch #1 Bulbasaur the Grass/Poison type pokemon",
        pokemon: {
            pokemon_id: "1",
            pokemon_name: "bulbasaur",
            pokemon_image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png",
            pokemon_type: "[\"Grass\",\"Poison\"]"
        },
        status: 0
    },
    {
        id: 10,
        name: "7",
        type: "pokemon",
        message: "Catch #7 Squirtle the Water type pokemon",
        pokemon: {
            pokemon_id: "7",
            pokemon_name: "squirtle",
            pokemon_image: "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/7.png",
            pokemon_type: "[\"Water\"]"
        },
        status: 0
    }
];


export const emptyStringInitialState = {
    value: ''
};

export const emptyArrayInitialState = {
    todos: []
};