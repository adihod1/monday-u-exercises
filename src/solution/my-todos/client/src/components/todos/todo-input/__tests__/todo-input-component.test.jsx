import renderer from "react-test-renderer";
import TodoInput from "../todo-input-component";
import {todosList} from "../../../../utils/test-utils";

test('renders correctly TodoItem ', () => {
    const tree = renderer.create(
        <TodoInput todo={todosList}/>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

