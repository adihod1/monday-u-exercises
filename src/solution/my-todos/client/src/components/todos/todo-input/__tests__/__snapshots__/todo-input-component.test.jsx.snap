// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo Input tests renders correctly TodoItem 2  1`] = `
<div
  className="todo-input"
>
  <input
    id="todo-input"
    onChange={[Function]}
    onKeyUp={[Function]}
    placeholder="Add your new todo"
    type="text"
    value=""
  />
  <button
    className=""
    id="todo-input-btn"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-plus "
      data-icon="plus"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M432 256c0 17.69-14.33 32.01-32 32.01H256v144c0 17.69-14.33 31.99-32 31.99s-32-14.3-32-31.99v-144H48c-17.67 0-32-14.32-32-32.01s14.33-31.99 32-31.99H192v-144c0-17.69 14.33-32.01 32-32.01s32 14.32 32 32.01v144h144C417.7 224 432 238.3 432 256z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </button>
</div>
`;

exports[`renders correctly TodoItem  1`] = `
<div
  className="todo-input"
>
  <input
    id="todo-input"
    onChange={[Function]}
    onKeyUp={[Function]}
    placeholder="Add your new todo"
    type="text"
    value=""
  />
  <button
    className=""
    id="todo-input-btn"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-plus "
      data-icon="plus"
      data-prefix="fas"
      focusable="false"
      role="img"
      style={Object {}}
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M432 256c0 17.69-14.33 32.01-32 32.01H256v144c0 17.69-14.33 31.99-32 31.99s-32-14.3-32-31.99v-144H48c-17.67 0-32-14.32-32-32.01s14.33-31.99 32-31.99H192v-144c0-17.69 14.33-32.01 32-32.01s32 14.32 32 32.01v144h144C417.7 224 432 238.3 432 256z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </button>
</div>
`;
